File: src/instance.cpp
#include "instance.hpp"

#include <set>
#include "utils/files.hpp"


Instance::Instance(const string& file_name) 
{
	json inst_jsn = get_json_file(file_name);

	this->prepare(inst_jsn);
	this->parse(inst_jsn);
	this->assign_arrays();
	this->assign_prev_ops();
}


void Instance::prepare(json inst_jsn)
{
	int ops_size = 0;
	int trains_size = 0;
	int succ_size = 0;
	int res_size = 0;
	int objs_size = 0;

	for (const json& train_jsn : inst_jsn["trains"]) {
		trains_size++;

		for (const json& op_jsn : train_jsn) {
			ops_size++;

			succ_size += op_jsn["successors"].size();
			if (op_jsn.contains("resources")) {
				res_size += op_jsn["resources"].size();
				for (const auto& res_jsn : op_jsn["resources"]) {
					this->add_res_name(res_jsn["resource"]);
				}
			}
		}
	}

	objs_size = inst_jsn["objective"].size();


	this->ops.reserve(ops_size);
	this->trains.reserve(trains_size);
	this->op_succ.reserve(succ_size);
	this->op_res.reserve(res_size);
	this->objs.reserve(objs_size);
}


void Instance::parse(json inst_jsn)
{
	set<int> res_set;
	int n_res_duplicates = 0;

	for (const json& train_jsn : inst_jsn["trains"]) {
		Train train;
		train.op_start = this->n_ops();

		for (const json& op_jsn : train_jsn) {
			Op op;
			
			op.train = this->n_trains();

			op.dur = op_jsn["min_duration"];
			
			if (op_jsn.contains("start_lb")) {
				op.start_lb = op_jsn["start_lb"];
			}

			if (op_jsn.contains("start_ub")) {
				op.start_ub = op_jsn["start_ub"];
			}

			for (int s : op_jsn["successors"]) {
				op.succ.size += 1;
				this->op_succ.push_back(s + train.op_start);
			}

			if (op_jsn.contains("resources")) {
				res_set.clear();

				for (const auto& res_jsn : op_jsn["resources"]) {
					Res res;

					res.idx = this->res_name_to_idx[res_jsn["resource"]];
					if (res_jsn.contains("release_time")) {
						res.time = res_jsn["release_time"];
					}

					if (res_set.find(res.idx) != res_set.end()) {
						n_res_duplicates += 1;
						continue;
					}
					res_set.insert(res.idx);

					op.res.size += 1;
					this->op_res.push_back(res);
				}
			}

			train.ops.size += 1;
			this->ops.push_back(op);
		}

		this->trains.push_back(train);
	}

	this->op_res.shrink_to_fit();

	vector<int> op_obj_idx(this->n_ops(), -1);

	for (const json& obj_jsn : inst_jsn["objective"]) {
		assert(obj_jsn["type"] == "op_delay");

		Obj obj;
		if (obj_jsn.contains("threshold")) {
			obj.threshold = obj_jsn["threshold"];
		}

		if (obj_jsn.contains("coeff")) {
			obj.coeff = obj_jsn["coeff"];
		}

		if (obj_jsn.contains("increment")) {
			obj.increment = obj_jsn["increment"];
		}

		if (obj.coeff == 0 && obj.increment == 0) {
			continue;
		}


		int train_i = obj_jsn["train"];
		int op_i = obj_jsn["operation"];

		int op_idx = this->trains[train_i].op_start + op_i;

		op_obj_idx[op_idx] = this->objs.size();
		this->objs.push_back(obj);
	}

	for (int o = 0; o < this->n_ops(); o++) {
		if (op_obj_idx[o] >= 0) {
			this->ops[o].obj = &(this->objs[op_obj_idx[o]]);
		}
	}

}

void Instance::assign_arrays()
{
	int ops_idx = 0;
	int op_succ_idx = 0;
	int op_res_idx = 0;

	for (Train& train : this->trains) {
		assert(train.op_start == ops_idx);
		train.ops.assign_ptr(this->ops, ops_idx);

		for (Op& op : train.ops) {
			op.succ.assign_ptr(this->op_succ, op_succ_idx);
			op.res.assign_ptr(this->op_res, op_res_idx);
		}
	}

	assert(ops_idx == this->n_ops());
	assert(op_succ_idx == this->n_op_succ());
	assert(op_res_idx == this->n_op_res());
}


void Instance::assign_prev_ops()
{
	this->op_prev.resize(this->n_op_succ());

	for (const Op& op : this->ops) {
		for (int s : op.succ) {
			this->ops[s].prev.size += 1;
		}
	}

	int idx = 0;
	for (Op& op : this->ops) {
		op.prev.assign_ptr(this->op_prev, idx);
	}
	assert(idx == this->n_op_prev());


	vector<int> prev_filled(this->n_ops(), 0);
	for (int o = 0; o < this->n_ops(); o++) {
		for (int s : this->ops[o].succ) {
			this->ops[s].prev[prev_filled[s]++] = o;
		}
	}
}



void Instance::add_res_name(string res_name)
{
	if (this->res_name_to_idx.find(res_name) == this->res_name_to_idx.end()) {
		this->res_name_to_idx[res_name] = this->n_res();
	}
}

File: python/instance.py
#!./venv/bin/python3
import sys
import json

from dataclasses import dataclass, field
from functools import cached_property

DEFAULT_INSTANCE = '../data/test/example_problem.json'

@dataclass
class Obj:
	threshold: int 	= field(default=0)
	coeff: int		= field(default=0)
	increment: int	= field(default=0)

@dataclass
class Op:
	train_idx: int		= field(default=-1)
	op_idx: int			= field(default=-1)
	fork_idx: int|None 	= field(default=None)
	
	dur: int			= field(default=0)
	start_lb: int		= field(default=0)
	start_ub: int|None	= field(default=None)
	
	succ: list[int]		= field(default_factory=list)
	res: list[int]		= field(default_factory=list)
	res_time: list[int]	= field(default_factory=list)

	obj: Obj|None		= field(default=None)

	@cached_property
	def n_succ(self):
		return len(self.succ)
	
	@cached_property
	def n_res(self):
		return len(self.res)
	
	@cached_property
	def idx(self):
		return (self.train_idx, self.op_idx)
	
	@cached_property
	def name(self):
		return f'({self.train_idx},{self.op_idx})'

class Instance:
	ops : list[list[Op]]
	forks : list[Op]
	res_idx : dict[str, int]
	total_dur : int

	def __init__(self, file_name):
		self.parse_json(file_name)

	def parse_json(self, file_name):
		self.ops = []
		self.res_idx = {}
		self.total_dur = 0

		with open(file_name, 'r') as fp:
			jsn = json.load(fp)

		for train_idx, train_jsn in enumerate(jsn['trains']):
			train_ops = []
			for op_idx, op_jsn in enumerate(train_jsn):
				op = Op(
					train_idx	= train_idx,
					op_idx	 	= op_idx,
					dur			= op_jsn['min_duration'],
					start_lb	= op_jsn.get('start_lb', 0),
					start_ub	= op_jsn.get('start_ub', None)
				)

				self.total_dur += op.dur

				for succ in op_jsn['successors']:
					op.succ.append(succ)

				if 'resources' in op_jsn:
					for res_jsn in op_jsn['resources']:
						res_name = res_jsn['resource']
						res_time = res_jsn.get('release_time', 0)
						
						if res_name in self.res_idx:
							res_idx = self.res_idx[res_name]
						else:
							res_idx = len(self.res_idx)
							self.res_idx[res_name] = res_idx

						op.res.append(res_idx)
						op.res_time.append(res_time)

					
				train_ops.append(op)

			self.ops.append(train_ops)

		self.n_ops = [len(t) for t in self.ops]
		self.start_ops = [t[0] for t in self.ops]
		
		for obj_jsn in jsn['objective']:
			if obj_jsn['type'] != 'op_delay':
				continue

			train_idx = obj_jsn['train']
			op_idx = obj_jsn['operation']
			
			obj = Obj()
			obj.threshold = obj_jsn.get('threshold', 0)
			obj.coeff 	  = obj_jsn.get('coeff', 0)
			obj.increment = obj_jsn.get('increment', 0)

			assert(obj.coeff == 0 or obj.increment == 0)

			self.ops[train_idx][op_idx].obj = obj

		self.forks = []
		for train in self.ops:
			for op in train:
				if op.n_succ > 1:
					op.fork_idx = len(self.forks)
					self.forks.append(op)


	def get_swaps(self):
		pass

	@cached_property
	def n_trains(self):
		return len(self.ops)
	
	@cached_property
	def n_res(self):
		return len(self.res_idx)
	
	@cached_property
	def n_forks(self):
		return len(self.forks)

if __name__ == '__main__':
	inst = Instance(sys.argv[1] if len(sys.argv) > 1 else DEFAULT_INSTANCE)
	print(inst.forks)


